cs4660/graph/graph.py

```diff
@@ -0,0 +1,269 @@
+"""
+
+from io import open
+from operator import itemgetter
+
+def construct_graph_from_file(graph, file_path):
+    """
+    TODO: read content from file_path, then add nodes and edges to graph object
+
+    note that graph object will be either of AdjacencyList, AdjacencyMatrix or ObjectOriented
+
+    In example, you will need to do something similar to following:
+
+    1. add number of nodes to graph first (first line)
+    2. for each following line (from second line to last line), add them as edge to graph
+    3. return the graph    """
+    
+    #sf = open(file_path, "r")
+    with open(file_path) as sf:
+        fil = sf.read().splitlines()
+
+    total_nodes = int(fil[0])
+
+    for i in range(total_nodes):
+        current_node = Node(i)
+        graph.add_node(current_node)
+
+    for line in range(1, len(fil)):
+        values = fil[line].split(":")
+       # edge = parse_line_to_edge(line)
+        edge = Edge(Node(int(values[0])),  Node(int(values[1])), int(values[2]))
+        graph.add_edge(edge)
+
+        sf.close()
+
+    return graph
+
+class Node(object):
+    """Node represents basic unit of graph"""
+    def __init__(self, data):
+        self.data = data
+
+    def __str__(self):
+        return 'Node({})'.format(self.data)
+    def __repr__(self):
+        return 'Node({})'.format(self.data)
+
+    def __eq__(self, other_node):
+        return self.data == other_node.data
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash(self.data)
+
+class Edge(object):
+    """Edge represents basic unit of graph connecting between two edges"""
+    def __init__(self, from_node, to_node, weight):
+        self.from_node = from_node
+        self.to_node = to_node
+        self.weight = weight
+    def __str__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.from_node, self.to_node, self.weight)
+    def __repr__(self):
+        return 'Edge(from {}, to {}, weight {})'.format(self.from_node, self.to_node, self.weight)
+
+    def __eq__(self, other_node):
+        return self.from_node == other_node.from_node and self.to_node == other_node.to_node and self.weight == other_node.weight
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return hash((self.from_node, self.to_node, self.weight))
+
+
+class AdjacencyList(object):
+    """
+    AdjacencyList is one of the graph representation which uses adjacency list to
+    store nodes and edges
+    """
+    def __init__(self):
+        # adjacencyList should be a dictonary of node to edges
+        self.adjacency_list = {}
+
+    def adjacent(self, node_1, node_2):
+        if node_1 in self.adjacency_list:
+            if node_2 in self.adjacency_list[node_1]:
+                 return True
+        return False
+
+    def neighbors(self, node):
+        rest = []
+        for edge in self.adjacency_list[node]:
+          rest.append(edge.to_node)
```

Comment:
Indentation is inconsistent. Did you just copy this line of code from somewhere else?